File: practice/mathematics/combinatorics/anti_palindromic_string/anti_palindromic_string.cpp
#include <iostream>

using namespace std;

const unsigned long long MOD = 1000000007;

unsigned long long pow_mod(unsigned long long i_m, unsigned long long i_n)
{
    unsigned long long res = 1;
    while (i_n > 0)
    {
        if (i_n & 0x1)
        {
            res *= i_m;
            if (res >= MOD)
            {
                res %= MOD;   
            }
        }
        
        i_m *= i_m;
        if (i_m >= MOD)
        {
            i_m %= MOD;   
        }
        
        i_n >>= 1;
    }
    return res;
}

int main()
{
    int t;
    unsigned long long n, m;
    
    cin >> t;
    while (t--)
    {
        cin >> n >> m;
        if (n == 1)
        {
            cout << m << endl;    
        }
        else if (n == 2)
        {
            cout << (m * (m - 1) % MOD) << endl;   
        }
        else
        {
            unsigned long long total = pow_mod(m - 2, n - 2);
            unsigned long long res = total * m;
            if (res > MOD)
            {
                res %= MOD;   
            }
            res *= (m - 1);
            cout << (res % MOD) << endl;   
        }
    }
    
    return 0;
}

File: practice/mathematics/combinatorics/anti_palindromic_string/anti_palindromic_string.hs
module Main where

import Control.Monad (forM_)

p :: Integer
p = 10 ^ 9 + 7

pow_mod :: Integer -> Integer -> Integer
pow_mod m n = pow_mod_util m n 1
    where
        pow_mod_util m n r
            | n == 0         = r
            | n `mod` 2 == 1 = pow_mod_util (m * m `mod` p) (n `div` 2) (r * m `mod` p)
            | otherwise      = pow_mod_util (m * m `mod` p) (n `div` 2) r

solve :: Integer -> Integer -> Integer
solve n m
    | n == 1    = m
    | n == 2    = m * (m - 1) `mod` p
    | otherwise = total * m * (m - 1) `mod` p
    where
        total = pow_mod (m - 2) (n - 2)

main :: IO ()
main = do
    t <- readLn :: IO Int
    forM_ [1..t] $ \_ -> do
        nm <- getLine
        let [n, m] = map (read :: String -> Integer) $ words nm
        print $ solve n m

